### ISSUE TITLE
bad_function_call during read operation

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylla#3491]
*Installation details*
Scylla version (or git commit hash): 2.2.rc1-0.20180516.eb646c61e
Cluster size: 4
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

During longevity of 2.2.rc1 the following "Exception when communicating with 172.30.0.45: bad_function_call" errors are thrown for about an hour.

That happened in 2 consecutive runs.
We can see that the errors are thrown for several shards (0, 5, 8) and almost always the same source host (172.30.0.229 in this case).

The errors appears while no nemesis is running (in both cases) and only c-s mixed and c-s read were running.

```
2018-06-04 05:11:42,728 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:11:42 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.229: bad_function_call
2018-06-04 05:12:15,948 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:12:15 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:12:17,346 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:12:17 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:12:21,503 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:12:21 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:24:18,432 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:24:18 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.229: bad_function_call
2018-06-04 05:29:16,955 remote           L0197 DEBUG| [172.30.0.138] [stdout] Jun 04 05:29:16 err    |  [shard 7] storage_proxy - Exception when communicating with 172.30.0.138: bad_function_call
2018-06-04 05:29:17,701 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:29:17 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.138: bad_function_call
2018-06-04 05:35:25,660 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:35:25 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.229: bad_function_call
2018-06-04 05:35:27,804 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:35:27 err    |  [shard 8] storage_proxy - Exception when communicating with 172.30.0.229: bad_function_call
2018-06-04 05:38:46,363 remote           L0197 DEBUG| [172.30.0.138] [stdout] Jun 04 05:38:46 err    |  [shard 7] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:39:18,276 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:39:18 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:11,816 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:11 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:11,890 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:11 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:14,426 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:14 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:18,175 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:18 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:25,332 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:25 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:36,785 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:36 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:40:46,469 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:40:46 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:41:04,122 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:41:04 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:41:54,193 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:41:54 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:41:57,853 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:41:57 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:23,191 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:23 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:28,227 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:28 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:37,633 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:37 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:38,498 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:38 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:52,193 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:52 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:42:54,167 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:42:54 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:43:11,763 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:43:11 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:43:27,224 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:43:27 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:43:33,639 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:43:33 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:43:39,126 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:43:39 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:44:01,274 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:44:01 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:44:04,593 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:44:04 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:44:53,797 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:44:53 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:44:57,145 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:44:57 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:44:57,246 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:44:57 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:45:12,369 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:45:12 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:45:48,664 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:45:48 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:46:27,495 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:46:27 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:46:31,916 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:46:31 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:46:47,406 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:46:47 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:47:02,751 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:47:02 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:47:04,029 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:47:04 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:47:09,265 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:47:09 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:47:52,809 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:47:52 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:48:31,333 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:48:31 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:48:53,309 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:48:53 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:48:59,274 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:48:59 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:49:06,149 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:49:06 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:49:48,270 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:49:48 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:49:52,424 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:49:52 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:50:02,240 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:50:02 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:50:06,556 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:50:06 err    |  [shard 0] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
2018-06-04 05:50:25,164 remote           L0197 DEBUG| [172.30.0.229] [stdout] Jun 04 05:50:25 err    |  [shard 5] storage_proxy - Exception when communicating with 172.30.0.45: bad_function_call
```

Looking at exceptions counter in prometheus we can see that the exceptions are correlated between source host and destination host.
![screencapture-54-210-24-44-9090-graph-2018-06-04-16_17_10](https://user-images.githubusercontent.com/20959584/40919691-e341ed92-6812-11e8-94ae-04a3cbbb6c92.png)

 We also can see that the errors are coming from read operations:
![screen shot 2018-06-04 at 16 21 20](https://user-images.githubusercontent.com/20959584/40919841-64c01952-6813-11e8-8dbb-2f634b902415.png)

**Comment by roydahan** [comment]
Testing now with "more verbose" rpms got from @gleb-cloudius trying to reproduce the issue.

**Comment by gleb-cloudius** [comment]
Exception source is here:
```
flat_mutation_reader transform<schema_upgrader>(flat_mutation_reader, schema_upgrader)::transforming_reader::fast_forward_to(position_range, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.hh:490
flat_mutation_reader::fast_forward_to(position_range, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.hh:381
 (inlined by) on_end_of_underlying_stream at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.cc:274
 (inlined by) operator() at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.cc:297
 (inlined by) apply at /usr/src/debug/scylla-2.2.rc2/seastar/core/apply.hh:36
 (inlined by) apply<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>::<lambda(bool)>, bool> at /usr/src/debug/scylla-2.2.rc2/seastar/core/apply.hh:44
 (inlined by) apply<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>::<lambda(bool)>, bool> at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1362
 (inlined by) then<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>::<lambda(bool)> > at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:932
 (inlined by) operator() at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.cc:301
do_void_futurize_apply<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>&> at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1319
 (inlined by) apply<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>&> at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1355
 (inlined by) do_until<make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()>, make_nonforwardable(flat_mutation_reader, bool)::reader::fill_buffer(seastar::lowres_clock::time_point)::<lambda()> > at /usr/src/debug/scylla-2.2.rc2/seastar/core/future-util.hh:503
 (inlined by) fill_buffer at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.cc:301
flat_mutation_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.hh:348
 (inlined by) seastar::future<bool> flat_mutation_reader::impl::fill_buffer_from<flat_mutation_reader>(flat_mutation_reader&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /usr/src/debug/scylla-2.2.rc2/flat_mutation_reader.cc:164
single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#3}::operator()() const at /usr/src/debug/scylla-2.2.rc2/row_cache.cc:433
 (inlined by) _ZN7seastar22do_void_futurize_applyIRZN34single_partition_populating_reader11fill_bufferENSt6chrono10time_pointINS_12lowres_clockENS2_8durationIlSt5ratioILl1ELl1000EEEEEEEUlvE1_JEEENSt9enable_ifIXsrNS_9is_futureIJNSt9result_ofIFT_DpOT0_EE4typeEEEE5valueENS_6futureIJEEEE4typeEOSF_SI_ at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1319
 (inlined by) seastar::future<> seastar::futurize<void>::apply<single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#3}&>(single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#3}&) at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1355
 (inlined by) seastar::future<> seastar::do_until<single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#3}, single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#2}>(single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#2}, single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#3}) at /usr/src/debug/scylla-2.2.rc2/seastar/core/future-util.hh:503
single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /usr/src/debug/scylla-2.2.rc2/row_cache.cc:438
single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}::operator()() const at /usr/src/debug/scylla-2.2.rc2/row_cache.cc:429
 (inlined by) seastar::apply_helper<single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply({lambda()#1}&&, std::tuple<>) at /usr/src/debug/scylla-2.2.rc2/seastar/core/apply.hh:36
 (inlined by) auto seastar::apply<single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}>(single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}&&, std::tuple<>&&) at /usr/src/debug/scylla-2.2.rc2/seastar/core/apply.hh:44
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::apply<single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}>(single_partition_populating_reader::fill_buffer(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda()#1}&&, std::tuple<>&&) at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:1362
 (inlined by) _ZZN7seastar6futureIJEE4thenIZN34single_partition_populating_reader11fill_bufferENSt6chrono10time_pointINS_12lowres_clockENS4_8durationIlSt5ratioILl1ELl1000EEEEEEEUlvE_S1_EET0_OT_ENUlSF_E_clINS_12future_stateIJEEEEEDaSF_ at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:943
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE4thenIZN34single_partition_populating_reader11fill_bufferENSt6chrono10time_pointINS_12lowres_clockENS5_8durationIlSt5ratioILl1ELl1000EEEEEEEUlvE_S2_EET0_OT_EUlSG_E_JEE15run_and_disposeEv at /usr/src/debug/scylla-2.2.rc2/seastar/core/future.hh:413
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.2.rc2/seastar/core/reactor.cc:2500
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.2.rc2/seastar/core/reactor.cc:2910
seastar::reactor::run_some_tasks() at /opt/scylladb/include/c++/7/chrono:377
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.2.rc2/seastar/core/reactor.cc:3057
seastar::smp::configure(boost::program_options::variables_map)::{lambda()#3}::operator()() const at /usr/src/debug/scylla-2.2.rc2/seastar/core/reactor.cc:4002
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.2.rc2/seastar/core/posix.cc:52
```

/cc @tgrabiec, @haaawk

**Comment by gleb-cloudius** [comment]
Looks like @denesb remove the exception throwing in the master

**Comment by avikivity** [comment]
@tgrabiec for our edification, can you explain the chain of events in the trace?

**Comment by haaawk** [comment]
It seems that we're trying to fast_forward a reader that was created as not fast forwarding. 
I suspect that the reader was created here: https://github.com/scylladb/scylla/blob/scylla-2.2.rc1/row_cache.cc#L364
Then we try to fast forward it here: https://github.com/scylladb/scylla/blob/scylla-2.2.rc1/flat_mutation_reader.cc#L274
Seems like a bug.

**Comment by haaawk** [comment]
I think the bug is here: https://github.com/scylladb/scylla/blob/scylla-2.2.rc1/row_cache.cc#L363
schema transforming reader does not allow fast_forwarding so when reader.fwd() == streamed_mutation::forwarding::no we can just return it as is. Otherwise we should wrap in into make_forwardable().

**Comment by haaawk** [comment]
I will send a patch soon.

**Comment by tgrabiec** [comment]
wt., 5 cze 2018 o 09:50 Piotr Jastrzębski <notifications@github.com>
napisał(a):

> I think the bug is here:
> https://github.com/scylladb/scylla/blob/scylla-2.2.rc1/row_cache.cc#L363
> schema transforming reader does not allow fast_forwarding so when
> reader.fwd() == streamed_mutation::forwarding::no we can just return it as
> is. Otherwise we should wrap in into make_forwardable().
>

That's exactly what we already do. What's the bug?

>

**Comment by haaawk** [comment]
I think that solves the bug.
[0001-Fix-read_directly_from_underlying.patch.txt](https://github.com/scylladb/scylla/files/2071381/0001-Fix-read_directly_from_underlying.patch.txt)

**Comment by haaawk** [comment]
@tgrabiec 
I think the situation is as follows:
1. make_delegating_reader creates a reader that is forwardable so when reader.fwd() == streamed_mutation::forwarding::no then we have to turn in into nonforwardable
2. If we have to transform schema on the way then we wrap res into a transforming reader which is nonforwardable so the situation is reversed. We can return it when reader.fwd() == streamed_mutation::forwarding::no but when it's set to streamed_mutation::forwarding::yes then we can't just return it and have to return make_forwardable(res) instead.

**Comment by tgrabiec** [comment]
The bug was that transforming reader didn't have support for fast
forwarding, Avi just committed a missing backport which fixes that.

It's better to implement fast forwarding in the transforming reader than to
wrap the whole thing into make_forwardable(), since the underlying reader
which is transformed already supports forwarding.


wt., 5 cze 2018 o 10:04 Piotr Jastrzębski <notifications@github.com>
napisał(a):

**Comment by haaawk** [comment]
Sure. It will work this way too.

**Comment by tgrabiec** [comment]
I'll explain the conditions leading to this, as requested by Avi.

TL;DR:
  - single-partition read misses in cache
  - there was a concurrent memtable flush of the same table
  - there was a concurrent schema change

The exception happens when the cache reader is filling its buffer. That
read is a single-partition read which missed in the cache (hence we have
single_partition_populating_reader). The reader found the partition in the
underlying mutation source but was not allowed to populate the cache,
because the cache was updated on memtable flush concurrently. In such case,
the reader will simply delegate to the underlying reader. The delegating
reader is created in read_directly_from_underlying(). We cannot simply
return the underlying reader, because the underlying reader is set up for
cache population: it is always in fast forwarding mode, and always uses
cache's current schema, regardless of the schema requested for the reading.
In this case, those schemas are different, which means that there was a
schema change concurrent with the read. The underlying reader is wrapped in
a transforming reader, which converts it to the desired schema.

Also, the cache reader was requested to be not in fast forwarding mode
(streamed_mutation::forwarding::no). Database reads on behalf of queries
are currently never fast forwarding. So we need to wrap the underlying
reader with a special reader which makes it conform to
streamed_mutation::forwarding::no. The cache's underlying reader always
conforms to streamed_mutation::forwarding::yes because that's required and
assumed by the cache populating reader. In fast forwarding mode, the reader
stops and generates end-of-stream for each clustering range. Initially, the
clustering range is just including the static row. So the un-forwarding
transformer, after consuming the static row, needs to fast forward the
underlying reader to a full clustering range, to make it produce the rest
of the partition. That fast forwarding call failed in this case, because
the transforming reader, with which we wrapped the underlying reader to
change the schema, didn't have fast_forward_to() implemented.

wt., 5 cze 2018 o 09:31 Gleb Natapov <notifications@github.com> napisał(a):
